libcontainer/container_linux.go

```diff
@@ -201,18 +201,30 @@ func (c *Container) Set(config configs.Config) error {
 
 // Start starts a process inside the container. Returns error if process fails
 // to start. You can track process lifecycle with passed Process structure.
-func (c *Container) Start(process *Process) error {
+func (c *Container) Start(process *Process) (retErr error) {
 	c.m.Lock()
 	defer c.m.Unlock()
+	defer func() {
+		if retErr != nil {
+			c.terminate(process)
+		}
+	}()
 	return c.start(process)
 }
 
 // Run immediately starts the process inside the container. Returns an error if
 // the process fails to start. It does not block waiting for the exec fifo
 // after start returns but opens the fifo after start returns.
-func (c *Container) Run(process *Process) error {
+func (c *Container) Run(process *Process) (retErr error) {
 	c.m.Lock()
 	defer c.m.Unlock()
+
+	defer func() {
+		if retErr != nil {
+			c.terminate(process)
+		}
+	}()
+
```

Comment:
Similar to https://github.com/opencontainers/runc/pull/4355/commits/150c32f624db41c882586c6ad09cc5b417d5f183#r1866835966, there's no need for a defer here. Something like this would work:

```go
	if !process.Init {
		return nil
	}
	err := c.exec()
	if err != nil {
		c.terminate(process)
	}
	return err
```